File: src/main/java/cz/kb/oleg/strategy/service/dispatcher/ValidatingStrategyDispatcher.java
package cz.kb.oleg.strategy.service.dispatcher;

import cz.kb.oleg.strategy.api.StrategyDispatcher;
import jakarta.validation.ConstraintViolationException;
import jakarta.validation.Validator;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class ValidatingStrategyDispatcher<T> implements StrategyDispatcher<T> {

    private final StrategyDispatcher<T> strategyDispatcherDelegate;
    private final Validator validator;

    public ValidatingStrategyDispatcher(StrategyDispatcher<T> strategyDispatcher, Validator validator) {
        this.strategyDispatcherDelegate = strategyDispatcher;
        this.validator = validator;
    }

    @Override
    public void dispatch(@NonNull T data) {
        var violations = validator.validate(data);
        if (!violations.isEmpty()) {
            throw new ConstraintViolationException(violations);
        }
        strategyDispatcherDelegate.dispatch(data);
    }

}

File: src/main/kotlin/com/winterhate/strategy/service/dispatcher/ValidatingStrategyDispatcher.kt
package com.winterhate.strategy.service.dispatcher

import com.winterhate.strategy.api.StrategyDispatcher
import jakarta.validation.ConstraintViolationException
import jakarta.validation.Validator

class ValidatingStrategyDispatcher<T>(
    private val strategyDispatcherDelegate: StrategyDispatcher<T>,
    private val validator: Validator
) : StrategyDispatcher<T> {

    override fun dispatch(data: T) {
        val violations = validator.validate(data)
        if (violations.isNotEmpty()) {
            throw ConstraintViolationException(violations)
        }
        strategyDispatcherDelegate.dispatch(data)
    }
}

